src/graph/graph.c

```diff
@@ -290,287 +124,291 @@ Graph *Graph_Get(RedisModuleCtx *ctx, RedisModuleString *graph_name) {
-            id_to_save --;
-            continue;
-        }
+    res = GxB_Matrix_Delete(M, dest_id, src_id);
+    assert(res == GrB_SUCCESS);
 
-        // Perform all necessary entity substitution.
-        if(entityStore == g->nodes) {
-            // Substitute entity within storage and adjacency and label matrices.
-            _Graph_ReplaceDeletedNode(g, zero, id_to_save, id_to_replace);
-        } else {
-            // Update edge id.
-            Edge *e = Graph_GetEdge(g, id_to_save);
-            e->id = id_to_replace;
-            DataBlock_CopyItem(entityStore, id_to_save, id_to_replace);
-        }
+    M = _Graph_GetRelationMap(g, r);
+    res = GxB_Matrix_Delete(M, dest_id, src_id);
+    assert(res == GrB_SUCCESS);
 
-        // A swap has been made, so we'll update our source and destination indices.
-        id_to_replace_idx ++;
-        if (id_to_replace_idx >= IDCount) break;
-        id_to_save --;
+    // See if source is connected to destination with additional edges.
+    bool connected = false;
+    int relationCount = Graph_RelationTypeCount(g);
+    for(int i = 0; i < relationCount; i++) {
+        M = Graph_GetRelation(g, i);
+        GrB_Matrix_extractElement_BOOL(&connected, M, dest_id, src_id);
+        if(connected) break;
     }
 
-    // Free all deleted entities from datablock
-    DataBlock_FreeTop(entityStore, IDCount);
-
-    GrB_Vector_free(&zero);
-}
-
-/* Accepts a *sorted* array of IDs for edges to be deleted. */
-void Graph_DeleteEdges(Graph *g, EdgeID *IDs, size_t IDCount) {
-    for(int i = 0; i < IDCount; i++) {
-        Edge *e = Graph_GetEdge(g, IDs[i]);
-        // Removes edge from graph hashtable and updates relevent matrices.
-        _Graph_DeleteEdge(g, e);
+    /* There are no additional edges connecting source to destination
+     * Remove edge from THE adjacency matrix. */
+    if(!connected) {
+        M = Graph_GetAdjacencyMatrix(g);
+        res = GxB_Matrix_Delete(M, dest_id, src_id);
     }
 
     // Remove edges from datablock.
-    _Graph_DeleteEntities(g, IDs, IDCount, g->edges);
-}
-
-/* Accepts a *sorted* array of IDs for nodes to be deleted.
- * The deletion is performed by swapping higher-ID nodes not scheduled
- * for deletion into lower vacant positions, until all IDs greater than
- * the updated node count are scheduled for deletion. The adjacency matrix
- * is then resized to remove these. */
-void Graph_DeleteNodes(Graph *g, NodeID *IDs, size_t IDCount) {
-    assert(g && IDs);
-    if(IDCount == 0) return;
-
-    /* Delete incoming/outgoing edges of deleted nodes;
-     * For every node marked for deletion, get a list of 
-     * both incoming and outgoing edges and delete thoese edges. */
-    Vector *edges = NewVector(Edge*, 32);
-    for(int i = 0; i < IDCount; i++) {
-        Node *n = Graph_GetNode(g, IDs[i]);
-        Graph_GetNodeEdges(g, n, edges, GRAPH_EDGE_DIR_BOTH, GRAPH_NO_RELATION);
-    }
+    DataBlock_DeleteItem(g->edges, ENTITY_GET_ID(e));
```

Comment:
Should free e->Entity properties before this call.

Reply:
👍